src/external/rlsw.h

```diff
@@ -0,0 +1,3425 @@
+    SWerrcode errCode;                                          // Last error code
+
+    sw_texture_t* loadedTextures;
+    int loadedTextureCount;
+
+    uint32_t* freeTextureIds;
+    int freeTextureIdCount;
+
+    uint32_t stateFlags;
+
+} sw_data_t;
+
+
+/* === Global Data === */
+
+static sw_data_t RLSW = { 0 };
+
+
+/* === Helper Functions === */
+
+static inline void sw_matrix_id(sw_matrix_t dst)
+{
+    dst[0]  = 1, dst[1]  = 0, dst[2]  = 0, dst[3]  = 0;
+    dst[4]  = 0, dst[5]  = 1, dst[6]  = 0, dst[7]  = 0;
+    dst[8]  = 0, dst[9]  = 0, dst[10] = 1, dst[11] = 0;
+    dst[12] = 0, dst[13] = 0, dst[14] = 0, dst[15] = 1;
+}
+
+static inline void sw_matrix_mul(sw_matrix_t dst, const sw_matrix_t left, const sw_matrix_t right)
+{
+    sw_matrix_t result;
+    for (int i = 0; i < 4; i++) {
+        for (int j = 0; j < 4; j++) {
+            float sum = 0.0;
+            for (int k = 0; k < 4; k++) {
+                sum += left[i * 4 + k] * right[k * 4 + j];
+            }
+            result[i * 4 + j] = sum;
+        }
+    }
+    for (int i = 0; i < 16; i++) {
+        dst[i] = result[i];
+    }
+}
+
+static inline void sw_vec4_transform(float dst[4], const float v[4], const sw_matrix_t mat)
+{
+    float tmp[4] = {
+        mat[0] * v[0] + mat[4] * v[1] + mat[8] * v[2] + mat[12] * v[3],
+        mat[1] * v[0] + mat[5] * v[1] + mat[9] * v[2] + mat[13] * v[3],
+        mat[2] * v[0] + mat[6] * v[1] + mat[10] * v[2] + mat[14] * v[3],
+        mat[3] * v[0] + mat[7] * v[1] + mat[11] * v[2] + mat[15] * v[3]
+    };
+
+    for (int i = 0; i < 4; i++) {
+        dst[i] = tmp[i];
+    }
+}
+
+static inline float sw_saturate(float x)
+{
+    // After several comparisons, this saturation method
+    // seems to be the most optimized by GCC and Clang,
+    // and it does not produce any conditional branching.
+
+    // However, it is possible that a clamp could be
+    // more efficient on certain platforms.
+    // Comparisons will need to be made.
+
+    // SEE: https://godbolt.org/z/5qYznK5zj
+
+    // Saturation from below: max(0, x)
+    float y = 0.5f * (x + fabsf(x));
+
+    // Saturation from above: min(1, y)
+    return y - 0.5f * ((y - 1.0f) + fabsf(y - 1.0f));
+
+    // return (x < 0.0f) ? 0.0f : ((x > 1.0f) ? 1.0f : x);
+}
+
+static inline float sw_lerp(float a, float b, float t)
+{
+    return a + t * (b - a);
+}
+
+static inline sw_vertex_t sw_lerp_vertex_PNTCH(const sw_vertex_t* a, const sw_vertex_t* b, float t)
+{
+    sw_vertex_t result;
+    for (int i = 0; i < offsetof(sw_vertex_t, screen) / sizeof(float); i++) {
+        ((float*)&result)[i] = sw_lerp(((float*)a)[i], ((float*)b)[i], t);
+    }
+    return result;
+}
+
+
+/* === Framebuffer Part === */
+
+static inline void sw_framebuffer_load(void** color, void** depth, int w, int h)
+{
+    int size = w * h;
+
+    *color = SW_MALLOC(SW_COLOR_PIXEL_SIZE * size);
+    *depth = SW_MALLOC(SW_DEPTH_PIXEL_SIZE * size);
```

Comment:
Question: Is there anything from the Image API we could use here? The Image API does support different pixel formats, along with creating the image itself. Likely not completely needed, but an interesting idea that may save some duplicate code.